### PR TITLE
feat(typings): add typings to support TypeScript

### PR DESCRIPTION
**Pull request by alan-agius4**
Add typings to support TypeScript

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,10 +16,12 @@
   },
   "devDependencies": {
     "should": ">= 0.0.1 <9.0.0",
-    "sinon": ">=1.17.1"
+    "sinon": ">=1.17.1",
+    "typescript": "^2.3.4"
```
Comment (michael-yx-wu) [inline package.json]: unlikely to make a different to this PR, but 2.4.1 is available now 😄

**Comment by alan-agius4** [comment]
With the carot it will take 2.4.1 anyways

**Comment by vanesyan** [comment]
Can you resolve the pr on top of master, please :)

**Comment by alan-agius4** [comment]
What do you mean by 'resolve' please? I am current away and ill be able to do it on Friday

**Comment by vanesyan** [comment]
Sorry, mean rebase, mistyped :)

**Comment by alan-agius4** [comment]
Ah okay, will do that on friday :)

**Comment by alan-agius4** [comment]
@vanesyan rebase done :)

File: package.json [hunk 2]
```diff
@@ -28,8 +23,12 @@
     "index.js"
   ],
   "dependencies": {
-    "@types/node": "^8.0.2",
-    "graceful-readlink": ">= 1.0.0"
+    "@types/node": "^8.0.2"
```
Comment (vanesyan) [inline package.json]: Shouldn't this go in peerDependencies?

**Comment by alan-agius4** [comment]
Can go, if you prefer it that way. Though its not neccessary.

Not a bog fan of peers unless it's neccessery.

I'll do it as you prefer though. Just let me know :)

**Comment by alan-agius4** [comment]
The simple reason why its dependency is that it that node is requeried by
commander.js because of the 'eventEmitter'

**Comment by vanesyan** [comment]
Just worry about people that don't need typing for commander, they'll get extra deps.

**Comment by alan-agius4** [comment]
Thats trie, but its a very small deps.

Though i can change it if you wish

**Comment by vanesyan** [comment]
I'm not too familiar with TypeScript, but what do other publishers do?

**Comment by alan-agius4** [comment]
Usefully they are set as dependencies even with the typyings packages
themselves.


https://www.npmjs.com/package/@types/commander

**Comment by vanesyan** [comment]
Sorry me, if it's a bit stupid question, but what do we want to include typings within package, when typings are already released in DefinitelyTyped? :)

**Comment by alan-agius4** [comment]
Having typyings in DefinitlyTyped is a 'last' resort for the community to
have typings of a library that the owner doesn't want to have within his
library. TypeScript recommand having them within the library and not within
DefinitlyTyped.

**Comment by vanesyan** [comment]
Sorry me, but I haven't found such suggestion by TypeScript. Furthermore as I understand they heavily rely on `@types` npm namespace, which from my understanding is referring to DefinitelyTyped.

**Comment by alan-agius4** [comment]
https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/declaration%20files/Publishing.md

**Comment by alan-agius4** [comment]
@types is indeed for DefinitlyTyped which is used only for non typescript
libaries, as typically typescript libraries always have typing inside the
library itself and now andays even js libraries have typings in them such
as 'moment'

Reason being is for maintability and no need for addition packages.

TypeScript doesn't realy on DefinitlyTyped at all, its the consumer that
realies on them if he needs non TypeScript ready libraries.

**Comment by vanesyan** [comment]
Ok, sorry me again, but previously I was able to use Node.js api out of the box when used typescript last time in the past. But sometimes not so far, I tried it again and now only way to use node.js api was to install 'em from DefinitelyTypes.

**Comment by michael-yx-wu** [comment]
Yes that's because node does not ship with typings by default. Despite adding an extra dependency that's unnecessary for JavaScript projects having a peer dependency of node typings 8+ will give typescript projects targeting node <8 build errors because of unsatisfied peer dependencies. 

Commander seems to work with node 7+ (possibly even earlier versions). I think we should use the typings of the minimum version of node that commander is compatible with and set as a peer dependency.

**Comment by michael-yx-wu** [comment]
Including node typings as a peer dependency is fine since it's a declaration that commander works with a minimum version of node, but you'll run into problems with:

0. projects not using typescript (npm / yarn will warn that the peer dependency is not met and will be annoying)
0. typescript projects relying on an older version of node: I believe commander works with at least node 7+ and possibly earlier versions?

I think we should keep it as an actual dependency to avoid `1.`. We should also use the actual minimum version of node typings that commander is compatible with although `2.` is not an immediate concern unless we change the typings dependency to a peer dependency.

**Comment by alan-agius4** [comment]
So, what do you guys want me to do?

**Comment by michael-yx-wu** [comment]
My preference is to keep it as a dependency but use the version that commander is minimally compatible with.

**Comment by vanesyan** [comment]
Thanks :+1:

**Comment by Marak** [comment]
Hrmm. 

By adding `@types/node` dependency, `commander` project is no longer zero-dependencies.

The total size of the installed code-base seems to be increased by over 10x.

Project was previously `57kb` now is `503kb`. That seems wrong to increase package size by 10x in order to support one specific framework ( TypeScript ).

Not sure what the project maintainers are thinking here. Please review this again @michael-yx-wu.

**Comment by xavdid** [comment]
There's a couple of gotchas happening here:

1. the `package.json` files array doesn't include the `index.d.ts`, so the typings aren't shipped to npm. TS complains when you try to import commander, since the types aren't installed. That's an easy fix. 
2. The more involved piece is the dependencies. The bundle size increase is probably mostly due to the Typescript devDep, which commander doesn't _really_ depend on. The more harmonious way to do this is to include the `index.d.ts` file, but not introduce any dependencies for it. You can include the `/// <reference types="node">` flag to access `EventEmitter` and anything else needed. 
3. Semi-related, after cloning, I couldn't get the type tests to fail. I changed a `.option` line to `.asdf` and tests still passed fine. That could super be a configuration issue on my end, but if they're not actually catching anything, they're adding complexity. It's easy enough to have maintainers edit the `index.d.ts` file whenever there are API changes (usually semver major & minor). 

So! That said, if I were to make a PR now, I'd remove deps, add a reference to the index, and edit files. That should get the bundle back down, because shipped files would go up by exactly 1 and we'd be back to the deps we had before. 

How does that sound?

**Comment by alan-agius4** [comment]
@xavdid the bundle size is still the same considering the fact that TypeScript is only for devDeps, to be honest though, finding solutions to keep the devDeps down by removing TS, for me it's not a good idea as like this you will remove the tests will keeps the sanity of the definition file.

Why is it such a problem to have an extra devDeps which is actual used? Just for an example would you consider removing `sinon`?

**Comment by xavdid** [comment]
I don't care about the bundle size, but other commenters seem to take issue with it. Since the API for a big project like this changes so infrequently, it's pretty easy to keep the typings up-to-date. I don't see a ton of value in the tests, but it's not a big deal. 

That's a good point about the tests, I'm still pretty new with commander. 

Nevertheless, the `files` issue remains.

**Comment by alan-agius4** [comment]
The files issue is fixed
https://github.com/tj/commander.js/pull/725#pullrequestreview-79401278

**Pull request merged**